recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py

```diff
@@ -0,0 +1,157 @@
+import argparse
+from bson.binary import Binary
+import datetime
+import difflib
+import mmap
+import os
+import pymongo
+import re
+import subprocess
+import sys
+
+class DB:
+    mongo_db_name = 'test'
+    mongo_db_collection_name = 'rtos_dmesg_logs'
+
+    def __init__(self, server, user, password):
+        mongo_client = pymongo.MongoClient("mongodb://{user}:{pwd}@{server}".format(
+            user=args.user, pwd=args.password, server=args.server
+            ),
+            w=1)
+        db = mongo_client[self.mongo_db_name]
+        self.dmesg_logs_collection = db[self.mongo_db_collection_name]
+
+    def insert(self, data):
+        return self.dmesg_logs_collection.insert_one(data)
+
+    def find(self, query):
+        return self.dmesg_logs_collection.find(query)
+
+def run_cmd(cmd):
+    output = subprocess.check_output(cmd).strip()
+    return output.decode('utf-8')
+
+def get_device_desc():
+    cmdoutput = run_cmd(['fw_printenv', 'DeviceDesc'])
+    return cmdoutput.split('=')[1]
+
+def get_kernel_version():
+    return run_cmd(['uname', '-r'])
+
+def get_architecture():
+    return run_cmd(['uname', '-m'])
+
+def get_os_version():
+    with open('/etc/os-release', 'rb') as file, mmap.mmap(file.fileno(), 0, access=mmap.ACCESS_READ) as mfile:
+        build_id = re.search(br'BUILD_ID=\"(.*)\"', mfile)
+        return build_id.group(1).decode()
+
+def is_next_kernel():
+    kernel_version = get_kernel_version()
+    return kernel_version.endswith('-next')
+
+def get_dmesg_log():
+    dmesg_log = run_cmd(['dmesg'])
+    return dmesg_log
+
+def upload_log(db, dmesg_log):
+    data = {}
+    data['kernel_version'] = get_kernel_version()
+    data['kernel_type'] = 'next' if is_next_kernel() else 'current'
+    data['device_desc'] = get_device_desc()
+    data['architecture'] = get_architecture()
+    data['os_version'] = get_os_version()
+    data['date'] = str(datetime.datetime.now())
+
+    header = ''
+    for key, val in data.items():
+        header = header + '# {}: {}\n'.format(key, val)
+
+    data['dmesg_log'] = header + dmesg_log
+
+    record = db.insert(data)
+    print('INFO: Uploaded current dmesg log from {} with _id {}'.format(data['date'], record.inserted_id))
+
+def strip_headers(dmesg_log):
+    output = ''
+    for line in dmesg_log.splitlines():
+        if not line.startswith('#'):
+            output = output + line + '\n'
+    return output
+
+def get_old_dmesg_log(db):
+    query = {}
+    query['kernel_version'] = {'$ne': get_kernel_version()}
+    query['kernel_type'] = 'next' if is_next_kernel() else 'current'
```

Comment:
In theory you should be able to compare against logs from '-next' kernels too. For example when we upgrade the current 5.15 kernel to the 'next' 6.0 one we should be able to use the '-next' logs already stored for 6.0 because of '-next' testing.

I'm also wondering if the query should be restricted to logs from the same 'major.minor' set as the current kernel. We might get failures just because dmesg output changes between major kernel versions.

Reply:
I'll try to make this more intelligent and pick existing `major.minor` log if it isn't too involved.

Don't the `major.minor` versions change every couple of months? I think it would still be valuable to diff against any old log if the one with same `major.minor` isn't available - the test will fail but we can still manually vet the diff from the test output.

Reply:
I see your point  - for 'next' it will change every couple of months, for the stable shipping kernel is less often but yes it would still be valuable to see the diff.

Reply:
Updated to prefer last log with same kernel major.minor version, but if not then it'll pick last log of same kernel type (current/next).